lib/layout/image_placeholder.dart

```diff
@@ -54,23 +54,48 @@ class FadeInImagePlaceholder extends StatelessWidget {
   /// See [Image.fit].
   final BoxFit fit;
 
+  @override
+  _FadeInImagePlaceholderState createState() => _FadeInImagePlaceholderState();
+}
+
+class _FadeInImagePlaceholderState extends State<FadeInImagePlaceholder> {
+  bool _isLoaded = false;
+
+  @override
+  void initState() {
+    super.initState();
+    // Once the image is loaded, we no longer need to show a placeholder
+    widget.image
+        .resolve(ImageConfiguration.empty)
+        .addListener(ImageStreamListener(
+      (imageInfo, synchronousCall) {
+        setState(() {
+          _isLoaded = true;
+        });
+      },
+    ));
+  }
+
   @override
   Widget build(BuildContext context) {
     return Image(
-      image: image,
-      excludeFromSemantics: excludeFromSemantics,
-      width: width,
-      height: height,
-      fit: fit,
+      image: widget.image,
+      excludeFromSemantics: widget.excludeFromSemantics,
+      width: widget.width,
+      height: widget.height,
+      fit: widget.fit,
       frameBuilder: (context, child, frame, wasSynchronouslyLoaded) {
         // We only need to animate the loading of images on the web. We can
         // also return early if the images are already in the cache.
         if (wasSynchronouslyLoaded || !kIsWeb) {
-          return this.child ?? child;
+          if (!_isLoaded) {
+            return widget.placeholder;
+          }
+          return widget.child ?? child;
```

Comment:
I think we can simplify this a lot and avoid the bug mentioned above where we resolve it for an incorrect configuration by removing the _isLoaded flag altogether and just changing this code to:

```suggestion
          if (frame == null) {
            return widget.placeholder;
          }
          return widget.child ?? child;
```

FadeInImagePlaceholder can be a StatelessWidget again after that.
Not sure if we want the transition from placeholder to real image to be animated here as well or not (I am leaning towards animation).

Reply:
If we want it animated, we could simply fo this:

```dart
        if (wasSynchronouslyLoaded) {
          return this.child ?? child;
        } else {
          return AnimatedSwitcher(
            duration: duration,
            child: frame != null ? this.child ?? child : placeholder,
          );
        }
```

(And refactor the widget to be stateless again, of course).

Reply:
I originally tried what you suggested in the most recent comment, and when I did, it didn't work -- it works fine now. 

I think what was happening is I was testing state restoration while testing this part of the app and was getting confused -- Every time the app is killed and restored, the app ends up in the state at which `flutter run` was called, not including any of the changes I made to the gallery code afterwards.